### PR TITLE
Loans with margin calls

### PR DESCRIPTION
**Pull request by szerintedmi**

### CODE REVIEW

File: contracts/LoanManager.sol [hunk 1]
```diff
@@ -72,12 +73,26 @@ contract LoanManager is Restricted, TokenReceiver {
         rates = _rates;
     }
 
+    /* @Deprecated: For compatibility with previous versions (without the margin parameter), e.g. old SB scripts */
     function addLoanProduct(uint32 term, uint32 discountRate, uint32 collateralRatio, uint minDisbursedAmount,
```
Comment (szerintedmi) [inline contracts/LoanManager.sol]: Do we keep this temporary only? No client code uses it only local migration scripts

File: contracts/LoanManager.sol [hunk 2]
```diff
@@ -39,14 +39,15 @@ contract LoanManager is Restricted, TokenReceiver {
         uint32 productId; // 3
         LoanState state; // 4
         uint40 maturity; // 5
+        uint marginRate; // 6
```
Comment (szerintedmi) [inline contracts/LoanManager.sol]: we don't need to store it in Loan because loanProducts are immutable (only enable/disable)

File: contracts/LoanManager.sol [hunk 3]
```diff
@@ -22,13 +22,13 @@ contract LoanManager is Restricted, TokenReceiver {
...
-                                //      in parts per million , ie. 10,000 = 1%
-        uint32 defaultingFeePt; // 4: % of collateral in parts per million , ie. 50,000 = 5%
-        bool isActive;          // 5
+        uint minDisbursedAmount;    // 0: minimum loanAmount, with decimals set in AugmintToken.decimals (i.e. token amount)
+        uint32 term;                // 1: term length (in seconds)
+        uint32 discountRate;        // 2: discountRate (in parts per million, i.e. 10,000 = 1%)
+        uint32 collateralRatio;     // 3: repayment (token amount) / collateral (token amount), (in parts per million, i.e. 10,000 = 1%)
+        uint32 defaultingFeePt;     // 4: % of repaymentAmount (in parts per million, i.e. 50,000 = 5%)
+        bool isActive;              // 5: flag to enable/disable product
+        uint32 marginRatio;         // 6: ... (in ppm), zero means no margin
```
Comment (szerintedmi) [inline contracts/LoanManager.sol]: thinking about better name. `marginCallRatio` ? `marginCallThresholdRatio` ?

File: contracts/Rates.sol [hunk 4]
```diff
@@ -1,10 +1,15 @@
 /*
- Generic symbol / WEI rates contract.
+ Generic token / ETH rates contract.
  only callable by trusted price oracles.
  Being regularly called by a price oracle
-    TODO: trustless/decentrilezed price Oracle
+ Note: symbol is always the "pegged symbol" not the token's own symbol!
+    TODO: trustless/decentralized price Oracle
```
Comment (szerintedmi) [inline contracts/Rates.sol]: we should remove these TODOs if you agree

File: package.json [hunk 5]
```diff
@@ -38,7 +38,7 @@
     "localchaindb:build": "yarn localchaindb:clean && ./scripts/runganache.sh --db ./localchaindb & yarn migrate && echo 'Migration done. Ganache db in ./localchaindb folder.'",
     "localchaindb:builddocker": "yarn localchaindb:build && yarn docker:build",
     "docker:build": "docker build . -t localdockerimage",
-    "docker:run": "docker rm ganache ; docker run --init --name ganache -p 8545:8545 localdockerimage --db ./dockerLocalchaindb --gasLimit 0x47D5DE --gasPrice 1000000000 --networkId 999 -m \"hello build tongue rack parade express shine salute glare rate spice stock\" & wait",
+    "docker:run": "docker rm ganache ; docker run --init --name ganache -p 8545:8545 localdockerimage --db ./dockerLocalchaindb --gasLimit 0xFFFFFF --gasPrice 1000000000 --networkId 999 -m \"hello build tongue rack parade express shine salute glare rate spice stock\" & wait",
```
Comment (szerintedmi) [inline package.json]: setting the gasLimit so high can cause troubles because we will only see if contract too big when we try to deploy to Rinkeby/mainnet. I would set ganache gasLimit to the lower of RInkeby/mainnet block limit (and even a bit lower as blocklimit can shrink too)

File: contracts/SB_scripts/rinkeby/Rinkeby_0007_interestRates.sol [hunk 6]
```diff
@@ -93,11 +93,8 @@ contract Rinkeby_0007_interestRates {
     function disableAllLockProducts(Locker target) internal {
         uint32 productCount = uint32(target.getLockProductCount());
         for (uint32 i = 0; i < productCount; i++) {
-            uint32 perTermInterest;
-            uint32 durationInSecs;
-            uint32 minimumLockAmount;
             bool isActive;
-            (perTermInterest, durationInSecs, minimumLockAmount, isActive) = target.lockProducts(i);
+            (, , , isActive) = target.lockProducts(i);
```
Comment (szerintedmi) [inline contracts/SB_scripts/rinkeby/Rinkeby_0007_interestRates.sol]: why are we modifying scripts which we already executed?

File: contracts/LoanManager.sol (hunk 2 above)
Comment (szerintedmi) [reply contracts/LoanManager.sol]: + `getLoanTuple` needs to be expanded (maybe we won't need to keep the old getter for compatibility as we just add one more item to the end of the tuple?)

File: contracts/LoanManager.sol (hunk 3 above)
Comment (szerintedmi) [reply contracts/LoanManager.sol]: `getLoanProducts` needs to be expanded (maybe we won't need to keep the old getter for compatibility as we just add one more item to the end of the tuple?)

File: contracts/LoanManager.sol [hunk 7]
```diff
@@ -22,13 +22,13 @@ contract LoanManager is Restricted, TokenReceiver {
...
-        uint32 term;            // 1
-        uint32 discountRate;    // 2: discountRate in parts per million , ie. 10,000 = 1%
-        uint32 collateralRatio; // 3: loan token amount / colleteral pegged ccy value
-                                //      in parts per million , ie. 10,000 = 1%
-        uint32 defaultingFeePt; // 4: % of collateral in parts per million , ie. 50,000 = 5%
-        bool isActive;          // 5
+        uint minDisbursedAmount;    // 0: minimum loanAmount, with decimals set in AugmintToken.decimals (i.e. token amount)
+        uint32 term;                // 1: term length (in seconds)
+        uint32 discountRate;        // 2: discountRate (in parts per million, i.e. 10,000 = 1%)
+        uint32 collateralRatio;     // 3: repayment (token amount) / collateral (token amount), (in parts per million, i.e. 10,000 = 1%)
```
Comment (szerintedmi) [inline contracts/LoanManager.sol]: assuming we will have the option to originate a loan with more collateral (or add extra collateral later) then we should rename this to `minCollateralRatio`

File: contracts/LoanManager.sol [hunk 8]
```diff
@@ -72,12 +73,26 @@ contract LoanManager is Restricted, TokenReceiver {
         rates = _rates;
     }
 
+    /* @Deprecated: For compatibility with previous versions (without the margin parameter), e.g. old SB scripts */
```
Comment (szerintedmi) [inline contracts/LoanManager.sol]: Why do we keep this? No client code uses it only local migration scripts so we could just remove

File: contracts/Rates.sol (hunk 4 above)
Comment (treer-euedge) [reply contracts/Rates.sol]: agree

File: contracts/SB_scripts/rinkeby/Rinkeby_0007_interestRates.sol (hunk 6 above)
Comment (treer-euedge) [reply contracts/SB_scripts/rinkeby/Rinkeby_0007_interestRates.sol]: We unfortunately had to modify one: only this one rinkeby script tried to be too smart, and read data from the contract (via automatic getters), so we had to modify this slightly to compile. The change does not impact functionality, only there is one more number we ignore. Also i removed all the unused variables, also a noop change. (We only care about the isActive flag, all other properties are ignored)

File: contracts/LoanManager.sol (hunk 8 above)
Comment (treer-euedge) [reply contracts/LoanManager.sol]: All SB scripts would have to be modified (not only local, but rinkeby and mainnet too). In general we are trying to avoid modifying old SB scripts, which were already run.
Also all loan related unit tests would have to be modified without this.

File: contracts/LoanManager.sol (hunk 2 above)
Comment (treer-euedge) [reply contracts/LoanManager.sol]: It was not intended to be the final name, as this feature is still uder development. "marginRate" simply means the "rate" where the "margin" is. 
I'm not against a better name, but "marginCallExchangeRate" feels much more confusing to me. 

Yes, getLoanTuple has to be extended to. There is still a lot of stuff to do.
No tests or Sb scripts use getLoanTuple at all, that is only for the frontend.

Comment (treer-euedge) [reply contracts/LoanManager.sol]: We have settled on "marginCallRate" with Charlie.

File: package.json (hunk 5 above)
Comment (treer-euedge) [reply package.json]: Block limit is not constant, currently it is around 8 million on mainnet, and 7 million on rinkeby. Apart from small fluctuations it seems to be constantly growing (slowly).
The previous value of 4.7 million caused serious problems on the local ganache test network, so we had to increase it. We did not yet think about the correct final value, just increase it to get this obstacle out of the way, so we can make progress with the actual feature.
Setting it slightly below the lower of the above actual blocklimits seems sensible, i will set it to 6.9 million then.

File: contracts/LoanManager.sol (hunk 8 above)
Comment (szerintedmi) [reply contracts/LoanManager.sol]: Ouch, I see. But I don't like the idea to carry fx forever because of SB scripts. Also compile time will increase as the number of SB scripts are growing. What if we archive old SB scripts to a folder which doesn't compile? 
(tests shouldn't be a reason to keep a deprecated fx :) I'm happy to do the monkey work if it helps )

File: contracts/SB_scripts/rinkeby/Rinkeby_0007_interestRates.sol (hunk 6 above)
Comment (szerintedmi) [reply contracts/SB_scripts/rinkeby/Rinkeby_0007_interestRates.sol]: I see. See my prev comment about potentially archiving old SB scripts.

File: contracts/LoanManager.sol (hunk 8 above)
Comment (treer-euedge) [reply contracts/LoanManager.sol]: Archiving old SB scripts seems the way to go, i agree. - either in a non-compiled folder, or even to another repository?
Test will be updated eventually to use the new version anyway, as we progress with tests for the new features.

Comment (phraktle) [reply contracts/LoanManager.sol]: I suggest we simply delete them from the repo. For historical purposes, there's always the git repo history.

File: contracts/LoanManager.sol (hunk 3 above)
Comment (treer-euedge) [reply contracts/LoanManager.sol]: done

File: contracts/Rates.sol [hunk 9]
```diff
@@ -35,7 +32,7 @@ contract Rates is Restricted {
     }
 
     function setMultipleRates(bytes32[] symbols, uint[] newRates) external restrict("RatesFeeder") {
-        require(symbols.length == newRates.length, "symobls and newRates lengths must be equal");
+        require(symbols.length == newRates.length, "symbols and newRates lengths must be equal");
```
Comment (szerintedmi) [inline contracts/Rates.sol]: Do we really want to redeploy Rates contract just for this change? No strong opinion about it just double checking.

File: contracts/SB_scripts/rinkeby/Rinkeby_0007_interestRates.sol (hunk 6 above)
Comment (szerintedmi) [reply contracts/SB_scripts/rinkeby/Rinkeby_0007_interestRates.sol]: Are we archiving SB scripts and rolling this change back?

File: contracts/LoanManager.sol [hunk 10]
```diff
@@ -19,34 +19,34 @@ import "./MonetarySupervisor.sol";
...
-                                //      in parts per million , ie. 10,000 = 1%
-        uint32 defaultingFeePt; // 4: % of collateral in parts per million , ie. 50,000 = 5%
-        bool isActive;          // 5
+        uint minDisbursedAmount;    // 0: minimum loanAmount, with decimals set in AugmintToken.decimals (i.e. token amount)
+        uint32 term;                // 1: term length (in seconds)
+        uint32 discountRate;        // 2: discountRate (in parts per million, i.e. 10,000 = 1%)
+        uint32 collateralRatio;     // 3: inverse of collateral ratio: [repayment value (in token) / collateral value (in token)] (in ppm). Note: this is actually the inverse of the commonly used "collateral ratio"! TODO: fix it
+        uint32 defaultingFeePt;     // 4: % of repaymentAmount (in parts per million, i.e. 50,000 = 5%)
+        bool isActive;              // 5: flag to enable/disable product
+        uint32 minCollateralRatio;  // 6: minimum collateral ratio: [collateral value (in token) / repayment value (in token)] (in ppm), defines the margin, zero means no margin. Note: this is _not_ an inverse like the above "collateralRatio", it is already stored properly!
```
Comment (szerintedmi) [inline contracts/LoanManager.sol]: Wasn't the plan to call the new prop `marginCallRate` and rename existing`collateralRatio` prop to `minColleteralRatio`? Not too fussy but it's confusing in this current form.
How about `minInitialCollateralRatio` and `marginCallColletaralRatio?

File: contracts/LoanManager.sol [hunk 11]
```diff
@@ -19,34 +19,34 @@ import "./MonetarySupervisor.sol";
...
-        uint32 term;            // 1
-        uint32 discountRate;    // 2: discountRate in parts per million , ie. 10,000 = 1%
-        uint32 collateralRatio; // 3: loan token amount / colleteral pegged ccy value
-                                //      in parts per million , ie. 10,000 = 1%
-        uint32 defaultingFeePt; // 4: % of collateral in parts per million , ie. 50,000 = 5%
-        bool isActive;          // 5
+        uint minDisbursedAmount;    // 0: minimum loanAmount, with decimals set in AugmintToken.decimals (i.e. token amount)
+        uint32 term;                // 1: term length (in seconds)
+        uint32 discountRate;        // 2: discountRate (in parts per million, i.e. 10,000 = 1%)
+        uint32 collateralRatio;     // 3: inverse of collateral ratio: [repayment value (in token) / collateral value (in token)] (in ppm). Note: this is actually the inverse of the commonly used "collateral ratio"! TODO: fix it
```
Comment (szerintedmi) [inline contracts/LoanManager.sol]: Would it a be a big deal to fix it now? Or Separate PR?

File: contracts/LoanManager.sol (hunk 8 above)
Comment (szerintedmi) [reply contracts/LoanManager.sol]: So are we deleting this deprecated fx?

File: contracts/LoanManager.sol [hunk 12]
```diff
@@ -274,11 +275,12 @@ contract LoanManager is Restricted, TokenReceiver {
...
 
-        LoanState loanState =
-                loan.state == LoanState.Open && now >= loan.maturity ? LoanState.Defaulted : loan.state;
+        // Add extra calculated data for convenience: marginCallRate, isDefaulted
+        uint marginCallRate = calculateMarginCallRate(product.minCollateralRatio, loan.repaymentAmount, loan.collateralAmount);
 
         result = [loanId, loan.collateralAmount, loan.repaymentAmount, uint(loan.borrower),
-                loan.productId, uint(loanState), loan.maturity, disbursementTime, loanAmount, interestAmount];
+            loan.productId, uint(loan.state), loan.maturity, disbursementTime, loanAmount, interestAmount,
+            marginCallRate, isDefaulted(loan, currentRate) ? 1 : 0];
```
Comment (szerintedmi) [inline contracts/LoanManager.sol]: I think we shouldn't mix `isDefaulted` and `isUnderMargin` those are two different things. 
I.e. If a loan is under margin then it can still recover if someone tops up collateral. While a defaulted loan (which might be collected or not) can never recover, it can only be collected.

File: contracts/LoanManager.sol [hunk 13]
```diff
@@ -125,6 +140,16 @@ contract LoanManager is Restricted, TokenReceiver {
         emit NewLoan(productId, loanId, msg.sender, msg.value, loanAmount, repaymentAmount, maturity);
     }
 
+    function addExtraCollateral(uint loanId) external payable {
+        require(loanId < loans.length, "invalid loanId");
+        LoanData storage loan = loans[loanId];
+        require(loan.state == LoanState.Open, "loan state must be Open");
```
Comment (szerintedmi) [inline contracts/LoanManager.sol]: Do we want to allow adding collateral to a loan which is already defaulted (i.e. not repaid until maturity). 
It doesn't do too much harm as the collection will refund any leftover. 
If that's the case then we should be explicit in a comment about it. Alternatively, add an extra require for maturity.

File: contracts/LoanManager.sol (hunk 10 above)
Comment (phraktle) [reply contracts/LoanManager.sol]: What’s formerly known as `collateralRatio` should really be `initialCollateralRatio` (and inverted to fit how the term is commonly used). An aside: isn’t this defined by the loan product?

The new one is aptly called the minimum (position can be liquidated below this collateralization ratio).

File: contracts/LoanManager.sol (hunk 11 above)
Comment (phraktle) [reply contracts/LoanManager.sol]: I suggest addressing this separately, the goal of this PR is to add the basic margin call in the least invasive way.

File: contracts/LoanManager.sol (hunk 12 above)
Comment (phraktle) [reply contracts/LoanManager.sol]: Suggestion: call this property (and the method) `canCollect` or `collectable`

File: contracts/Rates.sol (hunk 9 above)
Comment (phraktle) [reply contracts/Rates.sol]: No :)

Comment (treer-euedge) [reply contracts/Rates.sol]: certainly not

File: contracts/LoanManager.sol (hunk 8 above)
Comment (treer-euedge) [reply contracts/LoanManager.sol]: We can only delete this after we've archived all old sb scripts and updated all the tests, and so there are no dependencies to it, no failing tests, no compile errors. After that is all done: sure we will delete it.

File: contracts/SB_scripts/rinkeby/Rinkeby_0007_interestRates.sol (hunk 6 above)
Comment (treer-euedge) [reply contracts/SB_scripts/rinkeby/Rinkeby_0007_interestRates.sol]: Archiving SB scripts should be done on a separate branch - as it is not really related to the loan-margin feature, and merge to this feature branch via staging.  After that this file will not even exist in its current location, making this change completely irrelevant.

File: contracts/LoanManager.sol (hunk 12 above)
Comment (treer-euedge) [reply contracts/LoanManager.sol]: I've renamed it.

File: contracts/LoanManager.sol (hunk 13 above)
Comment (treer-euedge) [reply contracts/LoanManager.sol]: I've no strong opinion on that.

File: contracts/LoanManager.sol (hunk 10 above)
Comment (treer-euedge) [reply contracts/LoanManager.sol]: I've no strong opinion on this.

File: contracts/LoanManager.sol (hunk 11 above)
Comment (treer-euedge) [reply contracts/LoanManager.sol]: I'd really like to fix this before the deploy, but i'm ok with doing it in a separate PR.

File: contracts/LoanManager.sol (hunk 7 above)
Comment (treer-euedge) [reply contracts/LoanManager.sol]: that's the initial value, not the minimum. (also since its an inverse it would technically be a maximum)
we've already renamed the actual minimum (~margin) to minCollateralRatio

File: contracts/LoanManager.sol (hunk 12 above)
Comment (szerintedmi) [reply contracts/LoanManager.sol]: Would an additional, calculated prop such as `isExpired` or `isDefaulted` still make sense? 
It could be calculated on the client too  but a bit wordy, something like: 
`maturity < now && (status == Open || status == Collected))`
Also client side calculation would require "blocktime" to be "in sync" with the client clock. 
I'm just thinking to settle with the interface early to avoid multiple changes of `@augmint/js` interface changes.

File: contracts/LoanManager.sol [hunk 14]
```diff
@@ -304,7 +304,7 @@ contract LoanManager is Restricted, TokenReceiver {
         return minCollateralRatio > 0 && marginCallRate > 0 && currentRate < marginCallRate;
     }
 
-    function isDefaulted(LoanData storage loan, uint currentRate)
+    function isCollectable(LoanData storage loan, uint currentRate)
```
Comment (szerintedmi) [inline contracts/LoanManager.sol]: I think it's misleading if we return true for `isCollectable` for loans which are already collected. We should always return `false` if it's already collected.
 AND add an additional flag(s?) to indicate if it was collected because it wasn't repaid on time or it was collected because it went under margin.

Comment (treer-euedge) [reply contracts/LoanManager.sol]: good point on checking for `Open` status (`Collected` and `Repaid` are not "collectable" indeed) - i've added that to `isCollectable`

about the additional flag: i'm not sure is that is really useful. needs extra storage on blockchain if separate flag. or we could extend the status enum to have 2 kinds of `Collected`. also what if it is both expired and under margin, how to decide what was the "real reason" then?

File: contracts/LoanManager.sol (hunk 8 above)
Comment (treer-euedge) [reply contracts/LoanManager.sol]: done: old sb scripts were archived, above function was deleted.

File: contracts/SB_scripts/rinkeby/Rinkeby_0007_interestRates.sol (hunk 6 above)
Comment (treer-euedge) [reply contracts/SB_scripts/rinkeby/Rinkeby_0007_interestRates.sol]: done: we've archived the old scripts, rolled back above change

File: contracts/LoanManager.sol [hunk 15]
```diff
@@ -271,15 +271,14 @@ contract LoanManager is Restricted, TokenReceiver {
...
+        loanAmount = repaymentAmount.mul(product.discountRate).div(1e6);
         interestAmount = loanAmount > repaymentAmount ? 0 : repaymentAmount.sub(loanAmount);
     }
 
     // the token/ETH rate of the margin, under which the loan can be "margin called" (collected)
     function calculateMarginCallRate(uint32 minCollateralRatio, uint repaymentAmount, uint collateralAmount)
     internal pure returns (uint) {
-        // TODO: think about proper div rounding to use here
-        return uint(minCollateralRatio).mul(repaymentAmount).mul(1000000000000).div(collateralAmount);
+        return uint(minCollateralRatio).mul(repaymentAmount).mul(1e12).div(collateralAmount);
```
Comment (phraktle) [inline contracts/LoanManager.sol]: the constant expression `1e18 / 1e6' might be more explanatory than `1e12` (and should be calculated at compile time, so no runtime cost)
